File: app/src/main/java/com/example/android/popularmovies/PosterAdapter.java
package com.example.android.popularmovies;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.example.android.popularmovies.model.MovieModel;
import com.example.android.popularmovies.util.MovieUtil;
import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by msk-1196 on 6/18/17.
 */

public class PosterAdapter extends RecyclerView.Adapter<PosterAdapter.PosterViewHolder> {
    private List<MovieModel> movies;
    private final PosterAdapterOnClickHandler mClickHandler;

    public interface PosterAdapterOnClickHandler {
        void onClick(MovieModel movie);
    }

    public void setMovies(List<MovieModel> movies) {
        this.movies = movies;
        notifyDataSetChanged();
    }

    public PosterAdapter(PosterAdapterOnClickHandler mClickHandler) {
        this.mClickHandler = mClickHandler;
    }

    @Override
    public PosterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new PosterViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.grid_movie_item, parent, false));
    }

    @Override
    public void onBindViewHolder(PosterViewHolder holder, int position) {
        holder.bind(movies.get(position));
    }

    @Override
    public int getItemCount() {
        if (movies == null)
            return 0;
        else
            return movies.size();
    }

    class PosterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        @BindView(R.id.iv_poster) ImageView mPosterImageView;

        private PosterViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(this);
        }

        private void bind(MovieModel movie){
            if (movie.getPosterPath() != null)
                Picasso.with(itemView.getContext())
                        .load(MovieUtil.getFullPosterPath(movie.getPosterPath()))
                        .into(mPosterImageView);
        }

        @Override
        public void onClick(View view) {
            mClickHandler.onClick(movies.get(getAdapterPosition()));
        }
    }
}

File: app/src/main/java/com/example/android/popularmovies/PosterAdapter.kt
package com.example.android.popularmovies

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView

import com.example.android.popularmovies.model.MovieModel
import com.example.android.popularmovies.util.MovieUtil
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.grid_movie_item.view.*


/**
 * Created by msk-1196 on 6/18/17.
 */

class PosterAdapter(private val mClickHandler: PosterAdapterOnClickHandler) :
        RecyclerView.Adapter<PosterAdapter.PosterViewHolder>() {
    private var movies: List<MovieModel>? = null

    interface PosterAdapterOnClickHandler {
        fun onClick(movie: MovieModel)
    }

    fun setMovies(movies: List<MovieModel>) {
        this.movies = movies
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PosterViewHolder {
        return PosterViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.grid_movie_item, parent, false))
    }

    override fun onBindViewHolder(holder: PosterViewHolder, position: Int) {
        holder.bind(movies!![position])
    }

    override fun getItemCount(): Int {
        return if (movies == null)
            0
        else
            movies!!.size
    }

    inner class PosterViewHolder (itemView: View)
        : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private var mPosterImageView: ImageView = itemView.iv_poster

        init {
            itemView.setOnClickListener(this)
        }

        fun bind(movie: MovieModel) {
            if (movie.posterPath != null)
                Picasso.with(itemView.context)
                        .load(MovieUtil.getFullPosterPath(movie.posterPath))
                        .into(mPosterImageView)
        }

        override fun onClick(view: View) {
            mClickHandler.onClick(movies!![adapterPosition])
        }
    }
}
